force-app/main/default/classes/SOQL.cls

```diff
@@ -16,343 +16,463 @@
+        } catch(Exception e) {
+            return '';
+        }
+    }
+
+    /**
+     * Question 20
+     * Returns account with related opportunities
+     * @param accountId The Account Id
+     * @return Account The account with opportunities
+     */
+    public static Account getAccountAndOpportunitiesFromAccountId(Id accountId) {
+        return [SELECT Id, Name,
+                (SELECT Id, Name, Amount, CloseDate FROM Opportunities)
+                FROM Account
+                WHERE Id = :accountId
+                LIMIT 1];
+    }
+
+    /**
+     * Question 21
+     * Returns contacts matching account state
+     * @param accountId The Account Id
+     * @return List<Contact> List of filtered contacts
+     */
+    public static List<Contact> getAccountsWithContactsByState(Id accountId) {
+        Account acc = [SELECT BillingState
+                      FROM Account
+                      WHERE Id = :accountId];
+        return [SELECT Id, Name, MailingState
+                FROM Contact
+                WHERE MailingState = :acc.BillingState
+                LIMIT 50000];
+    }
+
+    /**
+     * Question 22
+     * Returns accounts matching lead email domains
+     * @param leadIds Set of Lead Ids
+     * @return List<Account> List of matching accounts
+     */
+    public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
+        Set<String> emailDomains = new Set<String>();
+
+        for(Lead ld : [SELECT Email FROM Lead WHERE Id IN :leadIds LIMIT 50000]) {
+            if(String.isNotBlank(ld.Email)) {
+                String[] parts = ld.Email.split('@');
+                if(parts.size() > 1) {
+                    emailDomains.add(parts[1]);
+                }
+            }
+        }
+
+        return [SELECT Id, Website
+                FROM Account
+                WHERE Website LIKE :'%' + String.join(new List<String>(emailDomains), '%') + '%'
+                LIMIT 50000];
+    }
+
+    /**
+     * Question 23
+     * Clones opportunity line items
+     * @param origOppId Original Opportunity Id
+     * @param cloneOppIds List of cloned Opportunity Ids
+     */
+    public static void cloneLineItems(Id origOppId, List<Id> cloneOppIds) {
+        List<OpportunityLineItem> originalItems = [
+            SELECT Quantity, PricebookEntryId, UnitPrice
+            FROM OpportunityLineItem
+            WHERE OpportunityId = :origOppId
+            LIMIT 50000
+        ];
+
+        List<OpportunityLineItem> newItems = new List<OpportunityLineItem>();
+        Integer randomAdd = (Integer)(Math.random() * 3) + 1;
+
+        for(Id cloneId : cloneOppIds) {
+            for(OpportunityLineItem oli : originalItems) {
+                OpportunityLineItem newItem = new OpportunityLineItem(
+                    OpportunityId = cloneId,
+                    Quantity = oli.Quantity + randomAdd,
+                    PricebookEntryId = oli.PricebookEntryId,
+                    UnitPrice = oli.UnitPrice
```

Comment:
For practice, instead of setting these fields individually, try to use the clone() method.